test/iterate.cc

```diff
@@ -0,0 +1,728 @@
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(5, *s());
+}
+
+TEST(Iterate, ArrayElemSumRvalue) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(std::array<int, 2>({5, 12}))
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+  EXPECT_EQ(2, array_.size());
+}
+
+TEST(Iterate, ArrayElemSumMove) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(std::move(array_))
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+  EXPECT_EQ(2, array_.size());
+}
+
+TEST(Iterate, ArrayMovableElemSumMove) {
+  std::array<std::string, 2> array_ = {"Hello", "World"};
+
+  auto s = [&]() {
+    return Iterate(std::move(array_))
+        | Loop<std::string>()
+              .context(std::string(""))
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ("HelloWorld", *s());
+  EXPECT_EQ(2, array_.size());
+  EXPECT_EQ("", array_[0]);
+  EXPECT_EQ("", array_[1]);
+}
+
+TEST(Iterate, CommonArrayElemSumPtr) {
+  int x[] = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(&x[0], &x[1] + 1)
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+}
+
+TEST(Iterate, CommonArrayElemSum) {
+  int x[] = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(x, 2)
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+}
+
+TEST(Iterate, VectorStringConcat) {
+  std::vector<std::string> v = {"Hello", "World", "!"};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | Loop<std::string>()
+              .context(std::string(""))
+              .body([](auto& str, auto& stream, auto&& value) {
```

Comment:
How about s/str/result/?

Reply:
What do you mean? Need some more clarifications